File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
def solution(genres, plays):
    music_dict = {}
    
    idx = 0
    for (g, p) in zip(genres, plays):
        if g in music_dict:
            music_dict[g]['count'] += p
            music_dict[g]['list'].append({'idx' : idx, 'p' : p})
        else:
            music_dict[g] = {'count': p, 'list': [{'idx' : idx, 'p' : p}]}
        idx += 1
    
    sorted_data = dict(sorted(music_dict.items(), key=lambda x: x[1]['count'], reverse=True))
    
    result = []
    
    for i in sorted_data:
        i_list = sorted(music_dict[i]['list'], key=lambda x:x['p'], reverse=True)
        result.append(i_list[0]['idx'])
        if len(i_list) > 1:
            result.append(i_list[1]['idx'])
        
    return result

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.js
function solution(genres, plays) {
    const dict = {}

    genres.forEach((g, i) => {
        if (dict[g]) {
            dict[g].push({index : i, count: plays[i]})
        } else {
            dict[g] = [{index : i, count: plays[i]}]
        }
    })
    
    const sortedGenres = Object.entries(dict).sort((a, b) => {
        const totalA = a[1].reduce((acc, curr) => acc + curr.count, 0);
        const totalB = b[1].reduce((acc, curr) => acc + curr.count, 0);
        return totalB - totalA;
    })
    
    const result = []
    sortedGenres.forEach(([genres, songs]) => {
        songs.sort((a, b) => b.count - a.count)
        result.push(...songs.slice(0, 2).map(song => song.index))
    })
    
    return result
}